src/screencast/wlr_screencast.c

```diff
@@ -313,6 +313,205 @@ static void wlr_init_xdg_outputs(struct xdpw_screencast_context *ctx) {
+		}
+		fclose(f);
+		readin = true;
+	} else {
+		close(p1[1]);
+	}
+
+
+	if (exec_chooser(chooser->cmd, readin, p1, p2)) {
+		f = fdopen(p2[0], "r");
+		if (f == 0) {
+			perror("fdopen pipe2");
+			logprint(ERROR, "Failed to create stream reading from pipe2");
+			close(p2[0]);
+			*output = NULL;
+			return true;
+		}
+
+		ssize_t nread = getline(&name, &namelength, f);
+		fclose(f);
+		if (nread < 0) {
+			perror("getline failed");
+			*output = NULL;
+			return true;
+		}
+
+		//Strip newline
+		char *p = strchr(name, '\n');
+		if (p != NULL) {
+			*p = '\0';
+		}
+
+		logprint(TRACE, "wlroots: output chooser %s selects output %s", chooser->cmd, name);
+		wl_list_for_each_safe(out, tmp, output_list, link) {
+			if (strcmp(out->name, name) == 0) {
+				free(name);
+				*output = out;
+				return true;
+			}
+		}
+		free(name);
+		*output = NULL;
+		return true;
+	} else {
+		close(p2[0]);
+		*output = NULL;
+		return false;
+	}
+}
+
+static struct xdpw_wlr_output *wlr_output_chooser_default(struct wl_list *output_list) {
+	logprint(DEBUG, "wlroots: output chooser called");
+	struct xdpw_output_chooser default_chooser[] = {
+		{XDPW_CHOOSER_SIMPLE, "slurp -f %o -o"},
+		{XDPW_CHOOSER_DMENU, "wofi -d -n"},
+		{XDPW_CHOOSER_DMENU, "bemenu"},
+	};
+
+	int N = sizeof(default_chooser)/sizeof(default_chooser[0]);
+	struct xdpw_wlr_output *output = NULL;
+	bool ret;
+	for (int i = 0; i<N; i++) {
+		ret = wlr_output_chooser(&default_chooser[i], output_list, &output);
+		if (!ret) {
+			logprint(DEBUG, "wlroots: output chooser %s not found. Trying next one.",
+					default_chooser[i].cmd);
+			continue;
+		}
+		if (output != NULL) {
+			logprint(DEBUG, "wlroots: output chooser selects %s", output->name);
+		} else {
+			logprint(DEBUG, "wlroots: output chooser canceled");
+		}
+		return output;
+	}
+	return xdpw_wlr_output_first(output_list);
+}
+
+struct xdpw_wlr_output *xdpw_wlr_output_chooser(struct xdpw_screencast_context *ctx) {
+	if (ctx->state->config->screencast_conf.chooser_type == XDPW_CHOOSER_DEFAULT) {
+		return wlr_output_chooser_default(&ctx->output_list);
+	} else if (ctx->state->config->screencast_conf.chooser_type == XDPW_CHOOSER_NONE) {
+		struct xdpw_wlr_output *output = NULL;
+		if (ctx->state->config->screencast_conf.output_name) {
+			output = xdpw_wlr_output_find_by_name(&ctx->output_list, ctx->state->config->screencast_conf.output_name);
+		} else {
+			output = xdpw_wlr_output_first(&ctx->output_list);
+		}
+		return output;
+	} else {
+		struct xdpw_wlr_output *output = NULL;
+		if (!ctx->state->config->screencast_conf.chooser_cmd) {
+			logprint(ERROR, "wlroots: no output chooser given");
+			return NULL;
+		}
+		struct xdpw_output_chooser chooser = {
+			ctx->state->config->screencast_conf.chooser_type,
+			ctx->state->config->screencast_conf.chooser_cmd
+		};
+		logprint(DEBUG, "wlroots: output chooser %s (%d)", chooser.cmd, chooser.type);
+		bool ret;
+		ret = wlr_output_chooser(&chooser, &ctx->output_list, &output);
```

Comment:
Style: can be on the same line as the variable declaration.

Reply:
check